File: backend/src/utils/mixSorter.ts
import { ITask, TaskPriority } from '../models/Task';

/**
 * Calculates a dynamic score for a task based on Priority, Deadline Urgency, and Scheduled Time.
 * Higher score = higher priority in the list.
 */
export const calculateTaskMixScore = (task: ITask): number => {
  if (task.isCompleted) {
    return -10000; // Push completed tasks to the end
  }

  const now = new Date().getTime();

  // 1. Priority Base Weight
  const priorityWeights: Record<TaskPriority, number> = {
    urgent: 1000,
    high: 600,
    medium: 300,
    low: 100,
  };

  const priorityScore = priorityWeights[task.priority] || 300;

  // 2. Deadline Urgency Score
  let deadlineScore = 0;
  if (task.deadline) {
    const deadlineTime = new Date(task.deadline).getTime();
    const hoursRemaining = (deadlineTime - now) / (1000 * 60 * 60);

    if (hoursRemaining < 0) {
      // Overdue! Critical urgency!
      const hoursOverdue = Math.abs(hoursRemaining);
      deadlineScore = 2000 + Math.min(hoursOverdue * 50, 1000);
    } else if (hoursRemaining <= 3) {
      // Due within 3 hours
      deadlineScore = 1500;
    } else if (hoursRemaining <= 12) {
      // Due within 12 hours
      deadlineScore = 1000;
    } else if (hoursRemaining <= 24) {
      // Due within 1 day
      deadlineScore = 600;
    } else if (hoursRemaining <= 72) {
      // Due within 3 days
      deadlineScore = 300;
    } else {
      // Future deadline
      deadlineScore = Math.max(100 - (hoursRemaining / 24) * 10, 0);
    }
  }

  // 3. Scheduled Date-Time Factor
  let timeScore = 0;
  if (task.dateTime) {
    const scheduledTime = new Date(task.dateTime).getTime();
    const diffHours = (now - scheduledTime) / (1000 * 60 * 60);

    if (diffHours >= 0) {
      // Scheduled time has arrived or passed (ready to work on)
      timeScore = 200 + Math.min(diffHours * 10, 300);
    } else {
      // Scheduled for the future
      timeScore = Math.max(0, 100 - Math.abs(diffHours) * 5);
    }
  }

  return priorityScore + deadlineScore + timeScore;
};

/**
 * Sorts array of tasks using the Smart Mix Sorting Algorithm.
 */
export const sortTasksByMixAlgorithm = <T extends ITask>(tasks: T[]): T[] => {
  return [...tasks].sort((a, b) => {
    const scoreA = calculateTaskMixScore(a);
    const scoreB = calculateTaskMixScore(b);
    return scoreB - scoreA; // Descending score order
  });
};

File: backend/app/mix_sorter.py
from datetime import datetime, timezone
from app.models import Task

def calculate_task_mix_score(task: Task) -> int:
    """
    Calculates a multi-factor dynamic urgency score combining Priority, Deadline Urgency, and Scheduled Time.
    """
    if task.is_completed:
        return -10000

    now = datetime.utcnow()

    # 1. Priority Base Weight
    priority_weights = {
        "urgent": 1000,
        "high": 600,
        "medium": 300,
        "low": 100,
    }
    priority_score = priority_weights.get(task.priority.lower(), 300)

    # 2. Deadline Urgency Score
    deadline_score = 0
    if task.deadline:
        deadline_time = task.deadline
        hours_remaining = (deadline_time - now).total_seconds() / 3600.0

        if hours_remaining < 0:
            # Overdue
            hours_overdue = abs(hours_remaining)
            deadline_score = 2000 + int(min(hours_overdue * 50, 1000))
        elif hours_remaining <= 3:
            deadline_score = 1500
        elif hours_remaining <= 12:
            deadline_score = 1000
        elif hours_remaining <= 24:
            deadline_score = 600
        elif hours_remaining <= 72:
            deadline_score = 300
        else:
            deadline_score = max(0, int(100 - (hours_remaining / 24.0) * 10))

    # 3. Scheduled Time Factor
    time_score = 0
    if task.date_time:
        diff_hours = (now - task.date_time).total_seconds() / 3600.0
        if diff_hours >= 0:
            time_score = 200 + int(min(diff_hours * 10, 300))
        else:
            time_score = max(0, int(100 - abs(diff_hours) * 5))

    return priority_score + deadline_score + time_score

def sort_tasks_by_mix(tasks: list[Task]) -> list[Task]:
    return sorted(tasks, key=lambda t: calculate_task_mix_score(t), reverse=True)
